// prepack/lib/singletons.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setCreate = setCreate;
exports.setEnvironment = setEnvironment;
exports.setFunctions = setFunctions;
exports.setLeak = setLeak;
exports.setMaterialize = setMaterialize;
exports.setJoin = setJoin;
exports.setPath = setPath;
exports.setPathConditions = setPathConditions;
exports.setProperties = setProperties;
exports.setTo = setTo;
exports.setWiden = setWiden;
exports.setConcretize = setConcretize;
exports.setUtils = setUtils;
exports.setDebugReproManager = setDebugReproManager;
exports.DebugReproManager = exports.Utils = exports.concretize = exports.Widen = exports.To = exports.Properties = exports.createPathConditions = exports.Path = exports.Join = exports.Materialize = exports.Leak = exports.Functions = exports.Environment = exports.Create = void 0;

/**
 * Copyright (c) 2017-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
let Create = null;
exports.Create = Create;
let Environment = null;
exports.Environment = Environment;
let Functions = null;
exports.Functions = Functions;
let Leak = null;
exports.Leak = Leak;
let Materialize = null;
exports.Materialize = Materialize;
let Join = null;
exports.Join = Join;
let Path = null;
exports.Path = Path;
let createPathConditions = null;
exports.createPathConditions = createPathConditions;
let Properties = null;
exports.Properties = Properties;
let To = null;
exports.To = To;
let Widen = null;
exports.Widen = Widen;
let concretize = null;
exports.concretize = concretize;
let Utils = null;
exports.Utils = Utils;
let DebugReproManager = null;
exports.DebugReproManager = DebugReproManager;

function setCreate(singleton) {
  exports.Create = Create = singleton;
}

function setEnvironment(singleton) {
  exports.Environment = Environment = singleton;
}

function setFunctions(singleton) {
  exports.Functions = Functions = singleton;
}

function setLeak(singleton) {
  exports.Leak = Leak = singleton;
}

function setMaterialize(singleton) {
  exports.Materialize = Materialize = singleton;
}

function setJoin(singleton) {
  exports.Join = Join = singleton;
}

function setPath(singleton) {
  exports.Path = Path = singleton;
}

function setPathConditions(f) {
  exports.createPathConditions = createPathConditions = f;
}

function setProperties(singleton) {
  exports.Properties = Properties = singleton;
}

function setTo(singleton) {
  exports.To = To = singleton;
}

function setWiden(singleton) {
  exports.Widen = Widen = singleton;
}

function setConcretize(singleton) {
  exports.concretize = concretize = singleton;
}

function setUtils(singleton) {
  exports.Utils = Utils = singleton;
}

function setDebugReproManager(singleton) {
  exports.DebugReproManager = DebugReproManager = singleton;
}
